roachtest: include learner snapshot error in isExpectedRelocateError whitelist

nvb opened the pull request:
Fixes #40359.

We occasionally see the error "failed to apply snapshot: raft group deleted" during a RELOCATE RANGE statement. This occurs because the newly added learner attempts to send a raft message to another node after it has been removed and then destroys itself in response to a ReplicaTooOldError. We have a similar case in TestLearnerAdminChangeReplicasRace.

The PR also makes a few changes to improve the error reporting both in this test and in general when a command fails due to a context cancellation.

Release justification: Testing only.

nvb committed:
roachtest: return any context error from execCmd

5296a27 made is so that we would return context.Deadline errors instead
of "signal killed" errors from execCmd, but did nothing for context.Canceled
errors. This commit extends the logic to include any context error.

This turns:
```
cluster.go:1762,kv.go:587,cluster.go:2118,errgroup.go:57: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1569387937-58-n9cpu8:9 -- ./workload run kv --tolerate-errors --concurrency=512 --duration=10m --read-percent=50 {pgurl:1-8} returned:
    stderr:

    stdout:
        1.7     16.8     31.5    104.9 read
        83.0s        0        23995.7        24307.2     10.5     39.8     62.9    268.4 write
        84.0s        0        22667.4        24309.9      1.5     16.8     31.5    142.6 read
        84.0s        0        22984.5        24291.3      8.1     31.5     50.3    121.6 write
        85.0s        0        23746.2        24303.3      1.6     15.7     30.4   3087.0 read
        85.0s        0        23809.2        24285.7     10.0     39.8     65.0   3087.0 write
        86.0s        0        23188.4        24290.4      1.6     18.9     33.6    100.7 read
        86.0s        0        22974.4        24270.5      8.1     37.7     56.6    125.8 write
        87.0s        0        23691.4        24283.6      1.5     15.7     30.4   4295.0 read
        87.0s        0        24026.2        24267.8      8.9     35.7     65.0   4160.7 write
        88.0s        0        27901.2        24325.0      1.6     15.2     28.3     67.1 read
        88.0s        0        27651.3        24306.5      8.9     48.2     79.7    142.6 write
    : signal: killed
```

into:
```
cluster.go:1764,kv.go:589,cluster.go:2120,errgroup.go:57: /Users/nathan/Go/src/github.com/cockroachdb/cockroach/bin/roachprod run nathan-1569463684-01-n9cpu8:9 -- ./workload run kv --tolerate-errors --concurrency=512 --duration=10m --read-percent=50 {pgurl:1-8} returned:
    stderr:

    stdout:
    .2     14.2     54.5     79.7    226.5 write
        87.0s        0        19466.4        20437.9      2.2     19.9     33.6    100.7 read
        87.0s        0        19512.4        20461.9     13.1     48.2     79.7    251.7 write
        88.0s        0        19306.5        20425.4      2.2     24.1     41.9   3221.2 read
        88.0s        0        19266.6        20448.7     14.2     58.7     96.5   3221.2 write
        89.0s        0        19422.7        20414.1      2.2     23.1     48.2    142.6 read
        89.0s        0        19457.7        20437.5     15.7     60.8     96.5    419.4 write
        90.0s        0        19961.1        20409.3      2.0     22.0     41.9    167.8 read
        90.0s        0        20185.8        20434.9     15.2     56.6     88.1    142.6 write
    _elapsed___errors__ops/sec(inst)___ops/sec(cum)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
        91.0s        0        21390.8        20419.9      2.2     17.8     33.6     96.5 read
        91.0s        0        21201.7        20443.1     14.7     52.4     75.5    121.6 write
    : context canceled
```

Release justification: Testing only.

Release note: None

nvb committed:
roachtest: don't swallow relocate/split errors in kv50/rangelookups

Before, these errors would cause the context of the load generator to
be canceled, which would result in a "signal killed" error instead of
the real error from the test. This commit fixes that.

Release justification: Testing only.

Release note: None

nvb committed:
roachtest: include learner snapshot error in isExpectedRelocateError whitelist

Fixes #40359.

We occasionally see the error "failed to apply snapshot: raft group deleted"
during a RELOCATE RANGE statement. This occurs because the newly added learner
attempts to send a raft message to another node after it has been removed and
then destroys itself in response to a ReplicaTooOldError. We have a similar
case in TestLearnerAdminChangeReplicasRace.

Release justification: Testing only.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/41106)
<!-- Reviewable:end -->

tbg reviewed (commented):
> Reviewed 1 of 1 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41106)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @nvanbenschoten)

___
*[pkg/cmd/roachtest/kv.go, line 576 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/41106#-Lpgqxzb4wCJFzj_c0gb:-Lpgqxzc4jDxaaD8jBCG:bfldl1y) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6ac2dc9360bc925f3687a5c48445cd39a94e8a7/pkg/cmd/roachtest/kv.go#L576)):*
> ```Go
> 			defer close(doneWorkload)
> 			cmd := fmt.Sprintf("./workload init kv --splits=1000 {pgurl:1}")
> 			if err := c.RunE(ctx, c.Node(nodes+1), cmd); err != nil {
> ```

Hmm, this old code should have worked well enough to surface the error. `t.Fatal` should tell the test runner about the actual error and then `runtime,Goexit()` which `monitor.Go` should've translated to `errGoexit`: https://github.com/cockroachdb/cockroach/blob/7df2a12d716db8a782922d8e611d05cb9f6d6b0a/pkg/cmd/roachtest/cluster.go#L2098-L2106

If none of this was happening, we should investigate more. I had played with this a bunch to make sure that we could use the `Goexit`-int methods everywhere without having to have complicated rules about returning errors when inside of a `Go()` (which most of the roachtest code is in).


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41106)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @nvanbenschoten)

___
*[pkg/cmd/roachtest/kv.go, line 576 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/41106#-Lpgqxzb4wCJFzj_c0gb:-Lph8MlZDoP8K7SkhCfO:b-n9v7f3) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6ac2dc9360bc925f3687a5c48445cd39a94e8a7/pkg/cmd/roachtest/kv.go#L576)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Hmm, this old code should have worked well enough to surface the error. `t.Fatal` should tell the test runner about the actual error and then `runtime,Goexit()` which `monitor.Go` should've translated to `errGoexit`: https://github.com/cockroachdb/cockroach/blob/7df2a12d716db8a782922d8e611d05cb9f6d6b0a/pkg/cmd/roachtest/cluster.go#L2098-L2106

If none of this was happening, we should investigate more. I had played with this a bunch to make sure that we could use the `Goexit`-int methods everywhere without having to have complicated rules about returning errors when inside of a `Go()` (which most of the roachtest code is in).
</blockquote></details>

We could even improve this by instead of resorting to `errGoexit` in `m.Go` we'd pull out the result of the earlier `t.Fatal` call (something like `err = somethingMagicalToGetTheMessagePassedToFatalEarlier(m.t)`), then your code change would (for all intents and purposes) a noop. Not suggesting you should do that.


<!-- Sent from Reviewable.io -->

nvb committed:
roachtest: remove errTestAlreadyFailed

This was never necessary. The important part of a30e376 was that
`!m.t.Failed()` was checked in `monitor.Wait`

Release justification: Testing only.

Release note: None

nvanbenschoten commented:
I looked into what's going on here. The issue is that running `cluster.Run` in a monitor's errgroup has a habit of hiding real test failures. It's often the case that a goroutine will return an error to the monitor's errgroup. This will cancel the context of all other goroutines. Any goroutine in a `cluster.Run` call will then fail, which will result in a `test.Fatal`. Only then would the monitor's errgroup be able to synchronize and return the correct error to `Wait`, but by then the test had already fatal-ed so the real error is lost.

I'm curious what you think the best way to fix this is. We can:
1. treat this as a known issue and do nothing past f6ac2dc and similar fixes when these issues arise
2. remove `cluster.Run` and only allow `cluster.RunE`. This is unfortunate because we currently call this function 117 times and in a lot of those cases, it actually makes sense
3. disallow calls to `cluster.Run` from closures passed to `monitor.Go` by looking at the stack for callers
4. panic instead of calling `test.Fatal` _only_ when in a `monitor.Go` closure and capture the panics error
5. find a way to call `test.Fatal` immediately when any goroutine returns an error to the errgroup and ignore the error that comes from `errgroup.Wait`
6. find something else to do?

tbg reviewed (approved):
I have to take another look at this error propagation, thanks for the write up. Let's get this PR in though.

nvanbenschoten commented:
bors r+

nvanbenschoten commented:
bors r+
